pints/residuals_diagnostics.py

```diff
@@ -0,0 +1,272 @@
+
+    This function can be used to analyze the results of either optimisation or
+    MCMC Bayesian inference. When multiple samples of the residuals are present
+    (corresponding to multiple MCMC samples), the plot illustrates the
+    distribution of autocorrelations across the MCMC samples. At each lag,
+    a point is drawn at the median autocorrelation, and a line is drawn giving
+    the percentile range of the posterior interval specified as an argument (by
+    default, the 2.5th to the 97.5th percentile).
+
+    When multiple outputs are present, one residuals plot will be generated for
+    each output.
+
+    When a significance level is provided, confidence bounds for the sample
+    autocorrelations under the assumption of IID residuals are drawn on the
+    plot. Many of the observed residuals autocorrelations falling outside these
+    bounds could imply evidence against the residuals being IID.
+
+    Under the assumption that the residuals of length :math:`n` are IID with
+    mean 0 and variance :math:`\sigma^2`, for large :math:`n` the residuals
+    sample autocorrelations are approximately IID Normal(mean=0, variance=1/n).
+    This result is proved in [1]_ (see Theorem 7.2.2 and Example 7.2.1).
+    Therefore, the confidence bounds can be calculated by :math:`\pm z^*
+    n^{-1/2}` for the appropriate critical value :math:`z^*`.
+
+    This function returns a ``matplotlib`` figure.
+
+    Parameters
+    ----------
+    parameters
+        The parameter values with shape ``(n_samples, n_parameters)``. When
+        passing a single best fit parameter vector, ``n_samples`` will be 1.
+    problem
+        The problem given by a :class:``pints.SingleOutputProblem`` or
+        :class:``pints.MultiOutputProblem``, with ``n_parameters`` greater than
+        or equal to the ``n_parameters`` of the `parameters`. Extra parameters
+        not found in the problem are ignored.
+    max_lag
+        Optional int value (default 10). The highest lag to plot.
+    thinning
+        Optional int value (greater than zero). If thinning is set to ``n``,
+        only every nth sample in parameters will be used. If set to ``None``
+        (default), some thinning will be applied so that about 200 samples will
+        be used.
+    significance_level
+        ``None`` or float value (default 0.05). When a significance level is
+        provided, dashed lines for the confidence interval corresponding to
+        that significance level are drawn on the plot. When ``None``, no lines
+        are drawn.
+    posterior_interval
+        Float value (default 0.95). When multiple samples of the parameter
+        values are provided, this gives the size of the credible region of the
+        posterior to plot.
+
+    References
+    ----------
+    .. [1] Brockwell, P. J., & Davis, R. A. (1991). Time series: Theory and
+           methods (2nd ed.). New York: Springer.
+    """
+    import matplotlib.pyplot as plt
+    import numpy as np
```

Comment:
Numpy and scipy can be imported at module level.

We avoid importing matplotlib at the module level because users sometimes need to configure matplotlib before it's first imported. Module-level imports get executed when pints itself (or in this case `pints.residuals_diagnostics`) is imported, which would means users could no longer configure matplotlib after they've imported pints. To avoid that, we stick the matplotlib import in the method, because that only gets run when the method is executed.

(There's no overhead difference by the way, imported modules are cached globally so writing an import in a method only does the hard work once, the 2nd time the `import` statement returns a reference)